types.go

```diff
@@ -0,0 +1,120 @@
+package microerror
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+)
+
+const kindUnknown = "unknown"
+
+type Error struct {
+	Desc string `json:"desc,omitempty"`
+	Docs string `json:"docs,omitempty"`
+	Kind string `json:"kind"`
+}
+
+func (e Error) Error() string {
+	return toStringCase(e.Kind)
+}
+
+type JSONError struct {
+	Error `json:",inline"`
+
+	Annotation string       `json:"annotation,omitempty"`
+	Stack      []StackEntry `json:"stack,omitempty"`
+}
+
+type StackEntry struct {
+	File string `json:"file"`
+	Line int    `json:"line"`
+}
+
+type annotatedError struct {
+	annotation string
+	underlying Error
+}
+
+func (e annotatedError) Error() string {
+	if e.annotation == "" {
+		return e.underlying.Error()
+	}
+	return e.underlying.Error() + ": " + e.annotation
+}
+
+func (e annotatedError) MarshalJSON() ([]byte, error) {
+	o := JSONError{
+		Error: e.underlying,
+
+		Annotation: e.annotation,
+	}
+
+	bytes, err := json.Marshal(o)
+	if err != nil {
+		return nil, fmt.Errorf("microerror.annotatedError.MarshalJSON: %w object=%#v", err, o)
+	}
+
+	return bytes, nil
+}
+
+func (e annotatedError) Unwrap() error {
+	return e.underlying
+}
+
+type stackedError struct {
+	stackEntry StackEntry
+	underlying error
+}
+
+func (e stackedError) Error() string {
+	return e.underlying.Error()
+}
+
+func (e stackedError) MarshalJSON() ([]byte, error) {
```

Comment:
This function unwraps all the `stackedError` and reconstructs the stack. then it tries to find `annotatedError` to find the custom error annotation and finally tries to find `Error` or just creates one from arbitrary `error`. Then it sets all the fields to `JSONError` and finally marshals it using standard `json.Marshal` call.